File: src/main/scala/llmTests/a201501a/java/gpt4/CoursesCalendarImpl.java
package llmTests.a201501a.java.gpt4;

import llmTests.a201501a.java.CoursesCalendar;
import llmTests.shared.java.Pair;
import java.util.*;

public class CoursesCalendarImpl implements CoursesCalendar {

	private Map<Pair<Day, Room>, TreeMap<Integer, String>> schedule;

	public CoursesCalendarImpl() {
		schedule = new HashMap<>();
	}

	@Override
	public List<Integer> possibleSlots() {
		return new ArrayList<>(Arrays.asList(9, 10, 11, 12, 14, 15, 16, 17));
	}

	@Override
	public void bookRoom(Day d, Room r, int start, int duration, String course) {
		if (duration <= 0 || start < 9 || start + duration - 1 > 17) {
			throw new IllegalStateException("Invalid starting time or duration");
		}
		Pair<Day, Room> pair = new Pair<>(d, r);
		TreeMap<Integer, String> daySchedule = schedule.get(pair);
		if (daySchedule == null) {
			daySchedule = new TreeMap<>();
		}
		for (int i = start; i < start + duration; i++) {
			if(i == 13) continue;
			if(daySchedule.get(i) != null) throw new IllegalStateException("Cannot book due to conflict");
			daySchedule.put(i, course);
		}
		schedule.put(pair, daySchedule);
	}

	@Override
	public Set<Pair<Integer, String>> dayRoomSlots(Day d, Room r) {
		Set<Pair<Integer,String>> bookings = new HashSet<>();
		TreeMap<Integer, String> daySchedule = schedule.get(new Pair<>(d, r));
		if(daySchedule != null) {
			for (Map.Entry<Integer, String> entry : daySchedule.entrySet()) {
				bookings.add(new Pair<>(entry.getKey(), entry.getValue()));
			}
		}
		return bookings;
	}

	@Override
	public Map<Pair<Day, Room>, Set<Integer>> courseSlots(String course) {
		Map<Pair<Day, Room>, Set<Integer>> bookings = new HashMap<>();
		for (Map.Entry<Pair<Day, Room>, TreeMap<Integer, String>> entry : schedule.entrySet()) {
			Set<Integer> set = new HashSet<>();
			for (Map.Entry<Integer, String> timeEntry : entry.getValue().entrySet()) {
				if(timeEntry.getValue().equals(course)) {
					set.add(timeEntry.getKey());
				}
			}
			if(set.size() > 0) {
				bookings.put(entry.getKey(), set);
			}
		}
		return bookings;
	}
}

File: src/main/scala/llmTests/a201501a/scala/gpt4/CoursesCalendarImpl.scala
package llmTests.a201501a.scala.gpt4

import llmTests.a201501a.scala.CoursesCalendar
import llmTests.a201501a.scala.CoursesCalendar.{Day, Room}
import llmTests.shared.java.Pair
import scala.collection.mutable
import scala.jdk.CollectionConverters.*

class CoursesCalendarImpl extends CoursesCalendar {

  private val timetable: mutable.Map[(Day, Room, Int), String] = mutable.Map.empty

  private val slots: List[Int] = List(9, 10, 11, 12, 14, 15, 16, 17)

  def possibleSlots(): java.util.List[Integer] = slots.map(java.lang.Integer.valueOf(_)).asJava

  def bookRoom(d: Day, r: Room, start: Int, duration: Int, course: String): Unit = {
    require(slots.contains(start), "Invalid start hour.")
    require(slots.contains(start + duration - 1), "Invalid end hour.")

    val hours = start until (start + duration)

    for (timeSlot <- hours) {
      if (timetable.contains((d, r, timeSlot))) throw new IllegalStateException("Cannot book room. Time slot is already occupied.")
    }

    for (timeSlot <- hours) {
      timetable += ((d, r, timeSlot) -> course)
    }
  }

  def dayRoomSlots(d: Day, r: Room): java.util.Set[Pair[Integer,String]] = {
    timetable.filter { case ((day, room, _), _) => day == d && room == r  }
      .map { case ((_, _, hour), course) => new Pair[Integer, String](java.lang.Integer.valueOf(hour), course) }.toSet.asJava
  }

  def courseSlots(course: String): java.util.Map[Pair[Day, Room], java.util.Set[Integer]] = {
    val courseMap = timetable.filter { case (_, c) => c == course }
    val result = mutable.Map.empty[Pair[Day, Room], Set[Integer]]

    for (((day, room, hour), _) <- courseMap) {
      val key = new Pair[Day, Room](day, room)
      result.get(key) match {
        case Some(set) => result.update(key, set + hour)
        case None => result += (key -> Set(hour))
      }
    }

    val mapJava = result.map { case (pair, set) => (pair, set.map(java.lang.Integer.valueOf(_)).asJava) }.toMap.asJava

    mapJava
  }
}